src/components/Ref/Ref.tsx

```diff
@@ -22,18 +23,22 @@ export default class Ref extends Component<RefProps> {
     children: PropTypes.element,
 
     /**
-     * Called when componentDidMount.
+     * Called when a child component will be mounted or updated.
      *
      * @param {HTMLElement} node - Referred node.
      */
-    innerRef: PropTypes.func,
+    innerRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
   }
 
   componentDidMount() {
-    _.invoke(this.props, 'innerRef', findDOMNode(this))
+    handleRef(this.props, 'innerRef', findDOMNode(this))
+  }
+
+  componentWillUnmount() {
+    handleRef(this.props, 'innerRef', null)
   }
 
   render() {
-    return this.props.children && Children.only(this.props.children)
+    return this.props.children && React.Children.only(this.props.children)
```

Comment:
why we introduce this restriction? An alternative approach would be to encapsulate the same thing as client may do to resolve this issue for mutliple children (given that its client _intent_ to get first rendered DOM element), like:

```jsx
render() {
  <>
     {this.props.children}
  </>
}
```

Reply:
> Verifies that children has only one child (a React element) and returns it. Otherwise this method throws an error.

This ideally matches `Ref` idea because it will return a ref to a single element.

Reply:
to me it seems to be a question of safer API - with alternative suggested approach we will **never** throw an exception and prevent client's tree from being rendered, while, at the same time, for all of the cases will preserve behavior client would expect. There are certain reasons while classic refs could be assigned to single component only, but even with the fact how this `ref` abstraction is implemented in a `Ref` component **we already violate some things related to ref semantics - for example, with this implementation reference to component instance is never returned, in contrast to the way ref is implemented for React.**

So, don't see any serious violations here, and thus I would suggest to follow the rule of introducing as safe abstraction as possible to the client - the one that won't break entire rendering process for the client.

Reply:
A `Children.only` will throw this exception for multiple children:

```
Invariant Violation: React.Children.only expected to receive a single React element child.
```

And it's safe API, API that doesn't allow to shoot the leg. Your proposal is not an improvement it allows non obvious and confusing behaviour. It changes the behaviour silently.

- I will pass a single element and I will receive a single ref
- I will pass multiple elements and I will receive only first ref 🤔

Reply:
Refs only work on a single component.  It is common practice for components that require a single child to use `React.Children.only` to enforce this.  Let's not support multiple children.

Reply:
@levithomason 
If we are talking about common practice (and, thus, user expectations), we should start with the fact that `Ref` component we are introducing, in contrast to React `ref`, have **absolutely** different semantics
- `Ref` captures the topmost (or the first from multiple topmost) DOM element from rendered component's tree
- React `ref` always captures component ref for composite components, and DOM node for string components (like `div`, `button`, etc).

Here, as you might see, semantics of these two has completely diverged - and, thus, it is not right to compare common practices of React `ref` and `Ref` afterwards.

If we would ask ourselves what is the closest React utility which functionality is mimicked by `Ref`, it would be easy to conclude that `Ref` serves just as wrapper for `ReactDOM.findDOMNode()` method - and, thus, we should rather care about client's expectations on how this method behaves, not React `ref`.

It turns out that `ReactDOM.findDOMNode()` won't fail in cases when multiple elements are rendered, in contrast to what we have now for `Ref` component:

```jsx
class Example extends React.Component {
  render() {
    return ['hello', 'hi'].map(text => <div>{text}</div>))
  }

  componentDidMount() {
    // returns first rendered <div />
    console.log('DOM node', ReactDOM.findDOMNode(this))
  }
}
```

Thus, from 'common practice' and 'what are user expectations' point of view, a behavior where no exception will be thrown (with first DOM node being captured) is what is really expected in this case.

------------
@layershifter 
> And it's safe API, API that doesn't allow to shoot the leg. 

Here is an example where this API will do that:
```jsx
render() {
  return <Ref ...>{this.props.renderItems}</Ref>
}
```

If, for some reason, `renderItems()` would (change to) return array of React components, the leg will be shoot. And in _all_ these cases client fix will be to ensure that `Fragment` wrapper is used:

```jsx
<Ref ...>
  <>
     {this.props.renderItems}
  </>
</Ref>
```

So, the question is **why we won't make the API safe at the first place then?**. Note that `findDOMNode()` which semantics is reflected in implementation for `Ref`, doesn't throw in cases of components array rendered.

--------------

This all leads to broader discussion related to terminology used for naming. Actually, the name `Ref` seems to be quite misleading, as there is no React `ref` involved in the component's logic - its logic is about capturing DOM nodes, not component refs (it is a different story around `innerRef` that is used as storage for captured DOM node). It seems that `FindDOMNode` would be proper name for this component, as it would reflect logic semantics in precise and correct way:

```jsx
<FindDOMNode setNode={node => ...}>
  <One />
  <NoThrowIfAnotherOne />
</FindDOMNode>
```

Reply:
`findDOMNode()` is deprecated in `StrictMode` and we can't rely on it, it also forces us to use `react-dom`.

We're planning to redesign this component to use `forwardRef` API in the most cases and use `findDOMNode()` only as a fallback. So, if React's will decide to remove `findDOMNode()` it will not issue for us.

Reply:
I should note that `Ref` component is used widely in our components, so we should be more accurate with usage of DOM specific APIs and behaviours.